packages/calypso-e2e/src/lib/pages/media-page.ts

```diff
@@ -0,0 +1,148 @@
+
+	// Editor view
+	imageEditorCanvas: '.image-editor__canvas-container',
+	imageEditorToolbarButton: '.image-editor__toolbar-button',
+	imageEditorResetButton: 'button[data-e2e-button="reset"]',
+	imageEditorCancelButton: 'button[data-e2e-button="cancel"]',
+};
+
+/**
+ * Represents an instance of the WPCOM Media library page.
+ *
+ * @augments {BaseContainer}
+ */
+export class MediaPage extends BaseContainer {
+	/**
+	 * Constructs an instance of the MediaPage object.
+	 *
+	 * @param {Page} page Underlying page on which interactions take place.
+	 */
+	constructor( page: Page ) {
+		super( page, selectors.gallery );
+	}
+
+	/**
+	 * Post-initialization steps.
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async _postInit(): Promise< void > {
+		await this.page.waitForLoadState( 'domcontentloaded' );
+	}
+
+	/**
+	 * Given a 1-indexed number `n`, click on the nth item in the media gallery.
+	 *
+	 * If the media gallery has been filtered (eg. Images only), this method will select the
+	 * `nth` item in the filtered gallery.
+	 *
+	 * @param {[key: string]: number } param0 Parameter object.
+	 * @param {number} param0.item nth media gallery item to be selected.
+	 * @returns {Promise<void>} No return value.
+	 * @throws {Error} If requested item could not be located in the gallery. Alternatively,
+	 * if the click action failed to select the gallery item.
+	 */
+	async clickOn( { item }: { item: number } ): Promise< void > {
+		// Playwright is able to select the nth matching item given a selector.
+		// See https://playwright.dev/docs/selectors#pick-n-th-match-from-the-query-result.
+		const element = await this.page.waitForSelector(
+			`:nth-match(${ selectors.items }, ${ item })`
+		);
+
+		if ( ! element ) {
+			throw new Error(
+				`Requested item number ${ item } is greater than number of items in gallery.`
+			);
+		}
+
+		await element.waitForElementState( 'visible' );
+		await element.click();
+		await element.waitForElementState( 'stable' );
+		const isSelected = await element.evaluate( ( node ) =>
+			node.classList.contains( 'is-selected' )
+		);
+		if ( ! isSelected ) {
+			throw new Error( `Failed to select requested item number ${ item }` );
+		}
+	}
+
+	/**
+	 * Clicks on the navigation tab.
+	 *
+	 * @param {[key: string]: string } param0 Parameter object.
+	 * @param {string} param0.name Name of the tab to click.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async clickTab( {
+		name,
+	}: {
+		name: 'All' | 'Images' | 'Documents' | 'Videos' | 'Audio';
+	} ): Promise< void > {
+		await this.page.waitForSelector( selectors.navtabs );
+		await this.page.click( `${ selectors.navtabs } span:has-text("${ name }")` );
+		// Wait for all placeholders to disappear.
+		// Alternatively, waiting for `networkidle` will achieve the same objective
+		// at the cost of much longer resolving time (~20s).
+		await this.page.waitForSelector( '.is-placeholder', { state: 'hidden' } );
```

Comment:
Is it possible that `.is-placeholder` might be too generic here? Potentially match some other or even multiple elements on a page.

Reply:
The `.is-placeholder` label only appears when the loading animation is occurring, so I believe this is sufficient. Green local + TeamCity runs show that it works well under real tests.

Ideally, I would wait until page animations are done, but Playwright does not have a `waitForAnimation` method, so that's out. 

I suspect that if I add another line below this and instruct the gallery listing to be `stable` then it should be a 100% guard.

Reply:
I've added the guard by checking `is gallery stable?`, and the execution logs show that this addition consumes ~30ms on my machine.

```
  pw:api <= page.click succeeded +0ms
  pw:api => page.waitForSelector started +0ms
  pw:api waiting for selector ".is-placeholder" to be hidden +4ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +5ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +171ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +409ms
  pw:api   selector did not resolve to any element +518ms
  pw:api <= page.waitForSelector succeeded +4ms
  pw:api => elementHandle.waitForElementState started +0ms
  pw:api   waiting for element to be stable +3ms
  pw:api <= elementHandle.waitForElementState succeeded +32ms
```

Note that Playwright is constantly polling and matching on the `.is-placeholder` class as long as the placeholder imagess are showing, until all placeholder images are gone.

Reply:
Thanks for checking the timing of these actions. Great to see exactly what's happening under the hood. I guess the waiting for the placeholder element to be gone would be sufficient here then! 👍